### PR TITLE
WSG-64-extract-terminology-into-ocl-compatible-format

### PR DESCRIPTION
**Pull request by xkmato**
To convert a data dictionary into OCL

To test this, please run from the repo root directory

```
# For csv
 who_l3_smart_tools/cli/terminology.py -e who_l3_smart_tool/tests/data/l2/test_dd.xlsx -o /home/kenneth/Code/DiGi/who_l3_smart_tool/tests/data/l2 

# for json

 who_l3_smart_tools/cli/terminology.py -e who_l3_smart_tool/tests/data/l2/test_dd.xlsx -o who_l3_smart_tool/tests/data/l2 -f json

# The output will be in who_l3_smart_tool/tests/data/l2

```

I added some questions on the Jira ticket. Those are not addressed in the PR as I need some clarity.

I also would incline towards csv rather than xls because csv offers a lot more flexibility, especially with larger files. so I have added support for xls that only converts the book into csvs. Let me know what you think about that.

I have also found that I have to add the current working directory to the python path in order to import from core without having to install the package.

```
# In who_l3_smart_tools/cli/terminology.py add the code below at line 6
sys.path.insert(0, os.getcwd())
```

### CODE REVIEW

**Comment by ibacher** [comment]
@xkmato Could you run `poetry lock` as it suggests? Also do you have an example of the output of running this tool, say against the `test_dd` file?

**Comment by xkmato** [comment]
> @xkmato Could you run `poetry lock` as it suggests? Also do you have an example of the output of running this tool, say against the `test_dd` file?

I have done that @ibacher 

To test this, please run from the repo root directory

```
# For csv
 who_l3_smart_tools/cli/terminology.py -e who_l3_smart_tool/tests/data/l2/test_dd.xlsx -o /home/kenneth/Code/DiGi/who_l3_smart_tool/tests/data/l2 

# for json

 who_l3_smart_tools/cli/terminology.py -e who_l3_smart_tool/tests/data/l2/test_dd.xlsx -o who_l3_smart_tool/tests/data/l2 -of json

# The output will be in who_l3_smart_tool/tests/data/l2
```

I added some questions on the Jira ticket. Those are not addressed in the PR as I need some clarity.

I also would incline towards csv rather than xls because csv offers a lot more flexibility, especially with larger files. so I have added support for xls that only converts the book into csvs. Let me know what you think about that.

I have also found that I have to add the current working directory to the python path in order to import from core without having to install the package. 

```
# In who_l3_smart_tools/cli/terminology.py add the code below at line 6
sys.path.insert(0, os.getcwd())

```

**Comment by xkmato** [comment]
@ibacher 
Do you know how I should handle ICD, LONIC, ICHI and ICF codes? Should I just treat these like any other extra column(This is what is currently happening)? The reason I ask is because I see `Internal Concept Mappings` and `External Concept Mappings` and stand-alone mappings that are not part of the concept in the OCL docs and I think this is where these should be but I am not sure how to convert the input into the mappings. 

Mappings have URLs for example and I don't know where to find URLs for those.

On Repositories and organizations: How do you get this information from the provided data? Are there any columns that I can use to deduce what the repository or organization is.

**Comment by ibacher** [comment]
> I also would incline towards csv rather than xls because csv offers a lot more flexibility, especially with larger files. so I have added support for xls that only converts the book into csvs. Let me know what you think about that.

The only inputs this tool ever needs to take is XLSX files, because that's what the WHO publishes, and the idea is we want to semi-automate going from the L2 (XLSX and PDF) directly to the L3 (Implementation Guide + OCL) as much as possible.

> Do you know how I should handle ICD, LONIC, ICHI and ICF codes? 

For now, skip them, because I'm not altogether certain they are right. However, if they were processed, they should be turned into [OCL Mappings](https://docs.openconceptlab.org/en/latest/oclapi/apireference/mappings.html), because they are mapping the concepts to standard terminologies.

> The reason I ask is because I see Internal Concept Mappings and External Concept Mappings and stand-alone mappings that are not part of the concept in the OCL docs

So they are mappings to reference terminologies. Usually the URL will be an OCL specific URL in the form:

```
/orgs/:org/sources/:source/concepts/:conceptId
```

We should use that at least where OCL has an appropriate source. I think "External Concept Mappings" correspond to what you're talking about in terms of "stand-alone". They are "External" in the sense that they refer to a reference terminology not available in OCL.

**Comment by xkmato** [comment]
@ibacher I have updated the PR to also include ValueSets. To test it yourself, you can run `who_l3_smart_tools/cli/terminology.py -e tests/data/l2/test_dd.xlsx -o test_output -v test_output/values_sets.json -of json` or use `-of csv` for csv output. The output should also include ValueSets.

I have attached the two files if you prefer to just review them. Let me know what you think especially about the structure. 
For valuesets: 
1. I don't have information about the system so I've left what was in the docs. 
2. Some values have empty codes eg country of birth. It did look right but I don't know.
[hiv_concepts.json](https://github.com/user-attachments/files/16100313/hiv_concepts.json)
[values_sets.json](https://github.com/user-attachments/files/16100314/values_sets.json)

File: who_l3_smart_tools/cli/terminology.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    hiv_terminology = HIVTerminology.from_excel(args.excel_file)
+    if args.output_format == "json":
+        hiv_terminology.process_concept_for_json(os.path.join(args.output_dir, "hiv_concepts.json"))
+    else:
+        hiv_terminology.process_concept_for_csv(os.path.join(args.output_dir, "hiv_concepts.csv"))
+    hiv_terminology.write_value_sets(args.values_set)
+
+
+if __name__ == "__main__":
+    sys.exit(main())
```
Comment (ibacher) [inline who_l3_smart_tools/cli/terminology.py:69]:
```suggestion
    sys.exit(main())

```

File: who_l3_smart_tools/core/terminology/terminology.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        elif self.converted_row.get("datatype").lower() == "coding":
+            self.value_set = {
+                "system": "https://fhir.staging.aws.openconceptlab.org",
+                "type": {"coding": [], "text": self.converted_row["name"]},
+                "value": self.converted_row["id"].replace(".", ""),
+            }
+        else:
+            self.value_set = None
+
+    def process_for_csv(self):
```
Comment (ibacher) [inline who_l3_smart_tools/core/terminology/terminology.py:88]: IMO, rather than operating on a shared object (`self.converted_row()`), this should probably just return the row value in the expected format.

File: who_l3_smart_tools/core/terminology/terminology.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            if key in self.schema.exclude_columns:
+                continue
+            if self.schema.include_columns and key not in self.schema.include_columns:
+                continue
+            self.converted_row["extras"][
+                self.schema.format_extras_for_json(key)
+            ] = value
+
+
+class ConceptTerminology:
```
Comment (ibacher) [inline who_l3_smart_tools/core/terminology/terminology.py:114]: Could we rename this? It's not clear to me what this represents? Does it generate a OCL Source or OCL Collection? Does it collect the data from the data dictionary? "ConceptTerminology" sounds like nonsense to my ears.

File: who_l3_smart_tools/core/terminology/terminology.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        Process the concept data and write it to a CSV file.
+
+        Args:
+            write_to_path (str): The path to write the CSV file.
+
+        """
+        with open(write_to_path, "w") as f:
+            writer = csv.DictWriter(f, fieldnames=self.output_csv_header)
+            writer.writeheader()
+            for concept_class, data_file in self.files.items():
```
Comment (ibacher) [inline who_l3_smart_tools/core/terminology/terminology.py:197]: It seems like we could refactor this class to hold a list of `ConceptRow`, separate out the data input from the data output (i.e., make them different functions)

File: who_l3_smart_tools/core/terminology/who/schema.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import slugify
+from who_l3_smart_tools.core.terminology.schema import ConceptSchema
+
+def dash_preserving_slugify(text):
```
Comment (ibacher) [inline who_l3_smart_tools/core/terminology/who/schema.py:4]: This probably belongs in some kind of utility class right?

File: who_l3_smart_tools/core/terminology/who/terminology.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    def __init__(self, files: list):
+        owner_id = "WHO-Smart-Guidelines"
+        super().__init__(owner_id, files, schema=HIVConceptSchema)
+
+    @classmethod
+    def from_excel(cls, file_path: str):
+        directory = os.path.dirname(file_path)
+        files = {}
+        excel_df = pd.read_excel(file_path, sheet_name=None)
+        for sheet_name, df in excel_df.items():
```
Comment (ibacher) [inline who_l3_smart_tools/core/terminology/who/terminology.py:30]: I don't really love dumping this into files. It's a lot of files for something that should just fit in memory. Is there a reason we can't just operate over the data frame as we do elsewhere in this project?

**Comment by ibacher** [comment]
> Some values have empty codes eg country of birth. It did look right but I don't know.

That's right. There are, IIRC, two valuesets that are intended to be implementation-defined, so they won't have any values.

**Comment by ibacher** [comment]
Couple of comments on the concepts:

* Too many things are ending up in the extras, I think. We do not, generally, need all of that. (Maybe require, validation conditions, and activity_id. Everything else there should be processed differently). The goal is to end up with something matching the [terminology resources here](https://i-tech-uw.github.io/smart-hiv/artifacts.html#4), not necessarily a comprehensive view of the spreadsheet.
* To be valid OCL resources, we need a resource type, i.e. `"Type": "Concept"` in JSON or `resource_type` field in the CSV.
* `retired` is a boolean, not a string value, so should be `false` not `"FALSE"`
* Names and descriptions in OCL are lists of structured objects not strings. So, e.g., in JSON a name is represented as `{ "name": "First Name", "locale": "en", "locale_preferred": true, "name_type": "Fully Specified" }` and supplied via the `names` property. The CSV version uses columns, e.g., `name`, `name_locale`, `name_locale_preferred`, `name_type`. Note that for our purposes, only the `name` or `description` attribute for each member should vary from the example provided.
* `concept_class` should always be `Misc` for now.

The ValueSets are already part of the IG. The thing I think we need for them is OCL Collections, although if we can create ValueSets from the stuff in the IG, maybe that's already done?

File: who_l3_smart_tools/core/terminology/terminology.py (hunk 3 above)
Comment (xkmato) [reply who_l3_smart_tools/core/terminology/terminology.py:114]: It generates [concepts](https://docs.openconceptlab.org/en/latest/oclapi/apireference/concepts.html). This is actually the core of what the entire PR is doing. I was under the impression that the main output for this task is to convert concepts from the input xls into an output similar to [this](https://docs.openconceptlab.org/en/latest/oclapi/apireference/bulkimporting.html#concepts). Do you have suggestions on what is more appropriate to call it?

File: who_l3_smart_tools/core/terminology/who/schema.py (hunk 5 above)
Comment (xkmato) [reply who_l3_smart_tools/core/terminology/who/schema.py:4]: It does ideally but I prefer to refactor when needed. If it is only used here then it should stay here until there is a need to share it.

File: who_l3_smart_tools/core/terminology/who/terminology.py (hunk 6 above)
Comment (xkmato) [reply who_l3_smart_tools/core/terminology/who/terminology.py:30]: Good point. This decision was because I initially thought this was going to support both csv and xls. I will refactor this to use pyxl on the actual document.

**Comment by xkmato** [comment]
> > Some values have empty codes eg country of birth. It did look right but I don't know.
> 
> That's right. There are, IIRC, two valuesets that are intended to be implementation-defined, so they won't have any values.

Am I representing them right in the PR?

File: who_l3_smart_tools/core/terminology/who/schema.py (hunk 5 above)
Comment (ibacher) [reply who_l3_smart_tools/core/terminology/who/schema.py:4]: Sorry, I phrased that badly. What I meant was we already have the `who_l3_smart_tools.utils` namespace in this package, and it seems like a more logical place for this function to go.

File: who_l3_smart_tools/core/terminology/terminology.py (hunk 3 above)
Comment (ibacher) [reply who_l3_smart_tools/core/terminology/terminology.py:114]: Maybe go with `ConceptResourceGenerator`... longer but more accurate. Yes, you're right about the task; I just think the "Terminology" word is throwing me, since it has a domain-specific meaning (a terminology is a set of concepts, what this generates is resources for creating a representation of a terminology or something along those lines).

File: who_l3_smart_tools/core/terminology/terminology.py (hunk 4 above)
Comment (ibacher) [reply who_l3_smart_tools/core/terminology/terminology.py:197]: So what I'm basing this on is that we seem to have some duplication in both `process_concept_for_csv()` and `process_concept_for_json()`. In both cases, we iterate through a list of files, process each file into a list of rows, create a `ConceptRow` object, convert the row into it's appropriate representation and then write that to a file.

I was suggesting breaking this up a bit into a multi-step process, i.e.:

In `__init__()`:
* Iterate through a list of files
* Process each file into a list of ConceptRows
* Store the processed ConceptRows as an attribute

In `process_for_XXX()`, for each row in the stored ConceptRows:
* Convert the row to its appropriate representation (JSON or CSV)
* Write that representation to the file

Basically what this means is instead of needing to create a bunch of temporary `ConceptRow`s on each write operation, we do the read operation once, store those results, and then can produce both outputs from a single read. This trades a little memory usage (the `ConceptRow`s are retained for the lifetime of the `ConceptTerminology` class) to avoid having to read from disk twice, which given that our data set here is ~2100 rows, seems like a reasonable trade-off.

Comment (xkmato) [reply who_l3_smart_tools/core/terminology/terminology.py:197]: I see what you mean. I tend to incline towards memory than IO when it comes to working with files and I think that I why that was the more intuitive approach for me.

I think your suggested approach is much neater and if you're confident that the upper limit for the input size will easily fit in memory then I will make the refactor to do it that way.

One question though: Do you think there is a use case in which someone will want both json and csv outputs? This is perhaps the biggest reason to store the processed rows in RAM so that we can write at once for both options or other formats.

**Comment by xkmato** [comment]
@ibacher The changes also include making the valuesets optional since we are not yet decided whether we want it here or not.

**Review by ibacher — APPROVED** [review]
LGTM. Thanks @xkmato!
